### PR TITLE
KAFKA-13989: Errors while evaluating connector type should return UNKNOWN

### PR DESCRIPTION
**Pull request by gharris1727**
Signed-off-by: Greg Harris <greg.harris@aiven.io>

https://issues.apache.org/jira/browse/KAFKA-13989

This method is used by the REST API in the `GET /connectors?expand=info,status` call.
In cases where a connector has been created and later the connector plugin is removed, this generated a 500 error and hid the status of all other connectors. This is undesirable behavior as an error with the packaging of a single plugin severely impacts the operational visibility of the whole cluster.

Instead, we should utilize the `UNKNOWN` enum value to indicate that it is currently unknown what type of connector that plugin is. This should prompt the user to further investigate the packaging issues with the plugin.

In addition to fixing the reported issue, also simplify some call-sites with null-checks inside of the connectorTypeForConfig method.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java
Comment (soarez) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java:711]: Should there be a debug or info level log with the exception?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java]: Do we even need this separate method anymore? It looks like it's only ever called in tests. Can we inline this logic into `connectorTypeForConfig` and tweak our tests to call that method instead?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java]: If we remove `connectorTypeForClass`, we can probably just rename this to `connectorType`.

Also, the Javadoc states that `connConfig` may not be null. Are there cases where we anticipate that it will now be null? If so, we should update the Javadoc accordingly; if not, we should remove the null check.

**Review by C0urante — COMMENTED** [review]
Thanks @gharris1727. I like the simplification of using `connectorTypeForConfig` over `connectorTypeForClass`.

I'm wondering if we can add more testing coverage, since it doesn't appear that we explicitly cover the case outlined in the ticket involving requests to `/connectors?expand=status`. Can we add tests where we invoke `Herder::connectorStatus` (and possibly `Herder::connectorInfo`, since that's used when the `expand` query parameter is set to `info`) on a now-removed connector?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java
Comment (gharris1727) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java]: After doing some archaeology on the two null checks that are replaced by this one:
* https://github.com/apache/kafka/pull/10822
* https://github.com/apache/kafka/pull/3812

It appears that the null check is required when the method uses information from both the status backing store and the config backing store. In some situations, the config can be null while there are statuses for the connector and/or tasks:
* After a connector is created, a worker reads the new statuses without reading the new configs.
* After a connector is deleted, a worker reads the (absence of) new configs without reading the cleared statuses.

I think these are relatively narrow windows of opportunity for an NPE, but they still exist. In the same spirit as the current change, it would be better to return UNKNOWN in these situations than throw an NPE.

I'll update the method name and Javadoc.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java:219]: Confirmed that this error reaches the connectorStatus call without the patch applied.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java]: I like these additional assertions on the accuracy of the `ConnectorStateInfo` returned from `herder::connectorStatus`, but aren't we testing a bit of an edge case here since we're simulating a status being present for the connector but not a config?

I think if we want to add these assertions we should also tweak the test to match the more common case where both a config and statuses can be found for the connector.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java]: Could we downgrade this to `WARN`? `ERROR` may be a bit too noisy, especially if there's a UI or some other monitoring tool that's repeatedly making requests to `GET /connectors?expand=status`.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java]: Nit:
```suggestion
     * @return the {@link ConnectorType} of the connector, or {@link ConnectorType#UNKNOWN} if an error occurs or the type cannot be determined
```
(Since we gracefully handle null input by returning `UNKNOWN`, we shouldn't imply that it's only returned if an error occurs.)

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerderTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerderTest.java]: While we're in the neighborhood?
```suggestion
        String[] methodNames = new String[]{"connectorType", "buildRestartPlan", "recordRestarting"};
```

**Review by C0urante — COMMENTED** [review]
Thanks @gharris1727, looking pretty good! Can we also add a case to `AbstractHerderTest` that touches on `AbstractHerder::connectorInfo` when the plugin is missing from the worker, since that method also attempts to [determine the type of the connector](https://github.com/apache/kafka/blob/f7304db6d2d3a2a55db7e945586aeb91843dbacb/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java#L283) and is used to [satisfy requests to `GET /connectors?expand=info`](https://github.com/apache/kafka/blob/f7304db6d2d3a2a55db7e945586aeb91843dbacb/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java#L126)?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java]: Feel free to add test cases that cover these edge cases as well, if you'd like; happy to spend review time on that, but won't block this PR on it.

Comment (gharris1727) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java]: I was just leaving the null config in-place since it was already there, but i agree that it's an edge case that shouldn't be the main test.

I've replaced the missing config with a barebones config, and mocked out the plugins() calls. I think the other unit tests verifying the other null behaviors are sufficient, I won't repeat them for both status and info callers.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java
Comment (C0urante) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java:110]: We can remove the `BogusSourceConnector` class entirely now that it's not used anywhere.

**Review by C0urante — COMMENTED** [review]
Thanks @gharris1727, one more small suggestion and then this should be good to go, pending CI results.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java
Comment (gharris1727) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java:110]: I also removed BogusSourceTask since its connector would have been missing.

**Review by C0urante — APPROVED** [review]
LGTM

**Pull request merged**
